### PR TITLE
Block Supports: Extend stabilization to common experimental block support flags

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Part of: 
- https://github.com/WordPress/gutenberg/issues/64312

Depends on:
- https://github.com/WordPress/gutenberg/pull/66918

## What?

Extends the approach to block support stabilization in #66918 to also stabilize shared/common experimental block support flags such as:
- `__experimentalSkipSerialization`
- `__experimentalDefaultControls`

## Why?

These experimental flags have been around a long time and are extensively used throughout core blocks. It's time to stabilize them to increase confidence for extenders and better document what is possible.

## How?

- Leverages the same block type args filter as #66918 and #63401.
- Updates the filter's algorithm to stabilize shared experimental keys
- Keeps `__experimentalSkipSerialization` within the filtered results due to `wp_should_skip_block_supports_serialization` in core only checking `__experimentalSkipSerialization`. It will have to stick around in the Gutenberg filter until 6.8 is the minimum supported version.

## Next Steps

- [x] ~Update uses of `__experimentalSkipSerialization` and `__experimentalDefaultControls` within Gutenberg's block supports~
- [x] Clean up and performance checks

### Future follow-ups

- [ ] Updating core blocks within block library to use the stabilized keys.

## Testing Instructions

__Warning: This is likely going to need considerable testing!__

1. Add a range of blocks to a post that use multiple types of block supports e.g. border, color, typography etc.
2. Check that all styles for these blocks display correctly when adjusted
3. Make sure that the block has appropriate controls available within Global Styles and the block inspector
4. Modify the block.json config for all the blocks to use varying combinations of experimental and stable flags.
5. Repeat steps 2 & 3 to ensure the newly modified config is reflected and styling continues to work as before
6. Add a filter to your theme that filters the block type arg in different ways and repeat the earlier checks again
7. Add multiple filters that would switch things up at each application of a filter.
8. Confirm the end result of config stabilizations and merges are as expected.

## Screenshots or screencast 

There should be no visual differences in blocks between this PR and trunk. Any found are regressions.

### CODE REVIEW

File: backport-changelog/6.8/7069.md [hunk 1]
```diff
@@ -2,3 +2,4 @@ https://github.com/WordPress/wordpress-develop/pull/7069
 
 * https://github.com/WordPress/gutenberg/pull/63401
 * https://github.com/WordPress/gutenberg/pull/66918
+* https://github.com/WordPress/gutenberg/pull/67018
```
Comment (aaronrobertshaw) [inline backport-changelog/6.8/7069.md:5]: The plan is to backport all the block support stabilization updates in a single core PR as the PHP changes there will be done within relevant classes rather than via filters like the Gutenberg PRs. 

A single core commit will make debugging and tracing changes much easier as well.

**Comment by aaronrobertshaw** [comment]
I've kept this PR in a draft status for the moment. It would benefit from some additional testing and depends upon https://github.com/WordPress/gutenberg/pull/66918 merging first anyhow.

File: packages/blocks/src/store/test/process-block-type.js [hunk 2]
```diff
@@ -88,79 +88,7 @@ describe( 'processBlockType', () => {
 				radius: true,
 				style: true,
 				width: true,
-				__experimentalDefaultControls: {
```
Comment (aaronrobertshaw) [inline packages/blocks/src/store/test/process-block-type.js:91]: There was quite a bit of overlap and duplication between some of the test cases. I've consolidated some tests to keep things a little more concise as well as grouped relevant tests for deprecations within their own describe block.

**Comment by aaronrobertshaw** [comment]
I've flagged this as ready for review although the plan is still to merge https://github.com/WordPress/gutenberg/pull/66918 first then rebase this accordingly.

For now, my initial testing looks promising. I've taken multiple blocks with various default controls and serialization configs for a spin and they appear to be behaving for me. Adding several filters for block type args to my test theme works as expected given known limitations (i.e. end result can't be 100% accurate when there are repeated modifications of both the experimental and stable flags in differing orders via filters).

**Comment by aaronrobertshaw** [comment]
https://github.com/WordPress/gutenberg/pull/66918 has been merged and this PR branch rebased on trunk. 

I've also tweaked the inline comments for this PR to be inline with [feedback](https://github.com/WordPress/gutenberg/pull/66918#discussion_r1849637378) received on #66918.

File: lib/compat/wordpress-6.8/blocks.php [hunk 3]
```diff
@@ -20,77 +20,121 @@ function gutenberg_stabilize_experimental_block_supports( $args ) {
...
 
 	$updated_supports = array();
 	foreach ( $args['supports'] as $support => $config ) {
-		// Add the support's config as is when it's not in need of stabilization.
-		if ( empty( $experimental_to_stable_keys[ $support ] ) ) {
+		/*
+		 * If this support config has already been stabilized, skip it.
+		 * A stable support key occurring after an experimental key, gets
+		 * stabilized then so that the two configs can be merged effectively.
+		 */
```
Comment (tellthemachines) [inline lib/compat/wordpress-6.8/blocks.php:46]: I'm not sure I understand this. Is it possible for the same support to be declared twice for the same block?

Otherwise, couldn't we check if the support is already inside the `$updated_supports` array instead of using this `$done` array for that purpose?

File: lib/compat/wordpress-6.8/blocks.php [hunk 4]
```diff
@@ -20,77 +20,121 @@ function gutenberg_stabilize_experimental_block_supports( $args ) {
...
 
-			// If there is no stabilized key present, use the experimental config as is.
-			if ( ! array_key_exists( $stabilized_key, $args['supports'] ) ) {
-				$updated_supports[ $stabilized_key ] = $config;
-				continue;
+				/*
+				 * The `__experimentalSkipSerialization` key needs to be kept until
+				 * WP 6.8 becomes the minimum supported version. This is due to the
+				 * core `wp_should_skip_block_supports_serialization` function only
+				 * checking for `__experimentalSkipSerialization` in earlier versions.
```
Comment (tellthemachines) [inline lib/compat/wordpress-6.8/blocks.php:87]: We could also temporarily override the core function with an updated gutenberg version of it that handles the stable key instead of preserving the experimental one here.

File: lib/compat/wordpress-6.8/blocks.php [hunk 5]
```diff
@@ -20,77 +20,121 @@ function gutenberg_stabilize_experimental_block_supports( $args ) {
...
-			 * We cannot determine with certainty which value was "last modified" so
-			 * the insertion order is the best guess. The extreme edge case of multiple
-			 * filters tweaking the same support property will become less over time as
-			 * extenders migrate existing blocks and plugins to stable keys.
-			 */
+		// Stabilize the config value.
+		$stable_config = is_array( $config ) ? $stabilize_config( $config, $stable_support_key ) : $config;
+
+		/*
+		 * When both experimental and stable configs are present, use the order they
```
Comment (tellthemachines) [inline lib/compat/wordpress-6.8/blocks.php]: It would be nice to have a bit of context here on why both configs should be present, as it's not clear - would you ever have both for the same support? That doesn't seem like something that _should_ happen.

**Review by tellthemachines — COMMENTED** [review]
This is working well in testing! I have mostly questions on the code at this point, trying to understand how it all works 😄

File: lib/compat/wordpress-6.8/blocks.php (hunk 4 above)
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php:87]: That was the first option I looked into however it would require a tonne of updates across the block support PHP functions now and then again once the stabilization was in the core in the minimum supported version for Gutenberg.

This approach (leaving the experimental prop for two releases) only requires code in this 6.8 compat file, which will naturally be removed when 6.8 is the minimum version supported. No need for multiple rounds of updates.  

There's also the use case of 3rd party blocks that skip serialization checking via `wp_should_skip_block_supports_serialization`.

File: lib/compat/wordpress-6.8/blocks.php (hunk 5 above)
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php]: Does the rest of the comment below not provide the context here?

You are correct that a single block defining its supports wouldn't have both the experimental and stable keys. It's when plugins start filtering the block type args to turn on and off supports, tweak application of styles via skip serialization, or make more opinionated decisions on what controls should be displayed by default, that we can end up with both the experimental and stable keys.

Given this, any suggestions on how to improve the inline comment for clarity without it becoming a thesis? 🙏

File: lib/compat/wordpress-6.8/blocks.php (hunk 3 above)
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php:46]: > Is it possible for the same support to be declared twice for the same block?

Yes. Plugins can filter block type args. 

If core blocks use the stable config but a plugin still filters block type args using the experimental flags we'll get both. Likewise, if a block still uses the experimental key and a plugin filters using the stable key, we'll get both again.

In the case that we have an experimental key based config when it gets stabilized we need to see if it should be merged with any other stable config (or perhaps use the stable config value if it isn't an array/object). So if the experimental key is reached before the stable, we need to stabilize the config under the stable key while processing the experimental config so that the stable merged result can be stored under the stable key without overwriting anything that might have been filtered separately there.

As the loop continues and we reach a stable config key, if it was already stabilized and merged when processing the earlier experimental config, there's no need to do it again.

Does that make sense?

File: lib/compat/wordpress-6.8/blocks.php (hunk 4 above)
Comment (tellthemachines) [reply lib/compat/wordpress-6.8/blocks.php:87]: Hmm yeah that makes sense, best leave it as is.

File: lib/compat/wordpress-6.8/blocks.php (hunk 3 above)
Comment (tellthemachines) [reply lib/compat/wordpress-6.8/blocks.php:46]: That does make sense yes! I wasn't considering that plugins will be using the exact same filter we're using here 😄

I'm guessing that when this goes in core it will become a method of `class WP_Block_Type` instead?

File: lib/compat/wordpress-6.8/blocks.php (hunk 5 above)
Comment (tellthemachines) [reply lib/compat/wordpress-6.8/blocks.php]: Hmm maybe something like "If a plugin overrides the support config with the `register_block_type_args` filter, both experimental and stable configs may be present. In that case, use the order..."?
That points to the reason we might find two configs so I think it helps (it helps me at least 😅 )

File: lib/compat/wordpress-6.8/blocks.php (hunk 3 above)
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php:46]: > I'm guessing that when this goes in core it will become a method of class WP_Block_Type instead?

That's correct and touched on in this [comment](https://github.com/WordPress/gutenberg/pull/67018#discussion_r1847969189) in this PR. There's also some more context in the original [typography support stabilization PR](https://github.com/WordPress/gutenberg/pull/63401) and it's [backport](https://github.com/WordPress/wordpress-develop/pull/7069). 

That [backport](https://github.com/WordPress/wordpress-develop/pull/7069) though is out of date and needs updating with the other recent stabilizations.

File: lib/compat/wordpress-6.8/blocks.php (hunk 5 above)
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php]: Thanks, I like it. I've worked that into the inline comments via https://github.com/WordPress/gutenberg/pull/67018/commits/f502d300ba098c3308c7ada4f7e3c81718300e4e

**Review by ramonjd — APPROVED** [review]
I went through the testing steps and the changes are working as described when I mixed up the `__experimental*` and stable support keys. 

Doc changes are 👍🏻 

I'm not sure what more feedback or testing you need - let us know what would help, but if it's good with @tellthemachines, then it also LGTM.

On a side note, do you think it's worth advertising on the Core blog that stabile keys are a thing, and, I assume should be preferred? Or is there more stabilization to come?

**Comment by aaronrobertshaw** [comment]
Thanks for the review @ramonjd 👍 

> I'm not sure what more feedback or testing you need - let us know what would help, but if it's good with @tellthemachines, then it also LGTM.

If we've had a few people go through the PR test instructions and we can't see any obvious gaps, I think we might be best served by getting this into Gutenberg and testing more in the wild. I believe it should be fine, known issues for extreme edge cases aside e.g. repeated mix-and-match filtering using varied experimental and stable keys.

> On a side note, do you think it's worth advertising on the Core blog that stabile keys are a thing, and, I assume should be preferred? Or is there more stabilization to come?

For sure. I was expecting we'd post on the [dev blog](https://developer.wordpress.org/news/) and provide a release dev note at least. I wanted to touch base with @ndiego, @justintadlock, and @bph, to tap their collective wisdom on the best way to communicate all this. Especially due to the edge cases if multiple plugins filter block supports with different keys.

In terms of more stabilization to come, I'm not 100% sure at this stage. 

The `__experimentalSelector` keys were "stabilized" in terms of the block should be using the [Block Selectors API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-selectors/). The Selectors API though was made to be backward compatible with the old experimental key. The difference there was that if the `selectors` API is present at all it takes precedence.

There appear to be quite a few other random experimental top level block support keys e.g. 
- `__experimentalOnMerge`,
- `__experimentalSlashInserter`
- `__experimentalToolbar`
- `__experimentalDisableBlockOverlay`, and more.

All of those are probably much more niche use cases and less likely to be leveraged by 3rd party devs. Their stabilization should be a simple addition to the stabilization maps added in this PR, when the time comes.

Long story short, for now I think the main goal is stabilizing the block supports used for design tools. This PR pretty much gets that done.

**Comment by ramonjd** [comment]
> If we've had a few people go through the PR test instructions and we can't see any obvious gaps, I think we might be best served by getting this into Gutenberg and testing more in the wild.

Sounds good to me. Great stuff.

**Comment by aaronrobertshaw** [comment]
I'll merge this and get to work in earnest on the combined backport for all the recent stabilizations.

**Comment by youknowriad** [comment]
Any plans on adding some documention about `defaultControls` to the Block API docs?

**Comment by aaronrobertshaw** [comment]
> Any plans on adding some documention about defaultControls to the Block API docs?

Absolutely. It's already in the works although I've paused publishing docs, block.json schema updates, and dev notes until the backport and now potentially the relocation of `defaultControls` settles down.

I'll update and close the loop here when they're published.

The tracking of block support stablization is covered by:
- https://github.com/WordPress/gutenberg/issues/64312
- https://github.com/WordPress/gutenberg/issues/63001

**Comment by aaronrobertshaw** [comment]
Quick update:

As discussed in https://github.com/WordPress/gutenberg/pull/67729#issuecomment-2553200639, the approach to block support stabilization needs more time so will likely be punted from 6.8. There is a revert PR available (https://github.com/WordPress/gutenberg/pull/68163) and anyone interested in stabilized block supports is advised to follow that for the latest 🙏
